Add support for attacher/detacher interface in Flex volume

chakri-nelluri opened the pull request:
Add support for attacher/detacher interface in Flex volume
This change breaks backward compatibility and requires to be release noted.

```release-note
Flex volume plugin is updated to support attach/detach interfaces. It broke backward compatibility. Please update your drivers and implement the new callouts. 
```

chakri-nelluri committed:
Add support for attacher/detacher interface in Flex volume

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41804)
<!-- Reviewable:end -->

chakri-nelluri commented:
@saad-ali @rootfs @MikaelCluseau PTAL

If the plugin is not installed on the master node(controller-manager to be exact), the infra is generating the following events.
```
 1m            1m              1       kubelet, 127.0.0.1                      Warning         FailedMount             Unable to mount volumes for pod "nginx-nfs_default(ad616cd7-f823-11e6-95e2-2c600c7bf1df)": timeout expired waiting for volumes to attach/mount for pod "default"/"nginx-nfs". list of unattached/unmounted volumes=[test]
  1m            1m              1       kubelet, 127.0.0.1                      Warning         FailedSync              Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "default"/"nginx-nfs". list of unattached/unmounted volumes=[test]
```

rootfs commented:
@k8s-bot ok to test

rootfs commented:
@kubernetes/sig-storage-pr-reviews

rootfs commented on pkg/volume/flexvolume/util.go:
```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017

rootfs commented on pkg/volume/flexvolume/util.go:
```diff
@@ -0,0 +1,111 @@
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+func addSecretsToOptions(options map[string]string, spec *volume.Spec, namespace string, host volume.VolumeHost) error {
```
you can reuse [GetSecretForPV](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/util.go#L133)

rootfs commented on pkg/volume/flexvolume/util.go:
```diff
@@ -0,0 +1,111 @@
+	return
+}
+
+func prepareForMount(mounter mount.Interface, deviceMountPath string) (alreadyMounted bool, err error) {
```
you used named return but `alreadyMounted` and `err` are not used.

rootfs commented on pkg/volume/flexvolume/util.go:
```diff
@@ -0,0 +1,111 @@
+// Mounts the device at the given path.
+// It is expected that prepareForMount has been called before.
+func doMount(mounter mount.Interface, devicePath, deviceMountPath, fsType string, options []string) error {
+	err := mounter.Mount(devicePath, deviceMountPath, fsType, options)
```
what if the device needs a format? Can `FormatAndMount` be called?

rootfs commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -0,0 +1,191 @@
+	plugin.unsupportedCommands = append(plugin.unsupportedCommands, commands...)
+}
+
+// Returns true iff the given command is know to be unsupported.
```
->known

rootfs commented on pkg/volume/flexvolume/driver-call.go:
```diff
@@ -0,0 +1,224 @@
+	// Option keys
+	optionFSType    = "kubernetes.io/fsType"
+	optionReadWrite = "kubernetes.io/readwrite"
+	optionKeySecret = "kubernetes.io/secret"
```
any selinux keys?

rootfs commented on pkg/volume/flexvolume/mounter.go:
```diff
@@ -0,0 +1,95 @@
+
+	call.AppendSpec(f.spec, f.plugin.host, extraOptions)
+
+	_, err = call.Run()
```
if `fsGroup` is present and !readOnly, call [SetVolumeOwnership](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gce_pd/gce_pd.go#L311)

chakri-nelluri commented:
Review status: 0 of 31 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/volume/flexvolume/driver-call.go, line 51 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxH-r1-51:-KdXERcP7wTWbdzadAO7:b-si323x) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/driver-call.go#L51)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary><blockquote>

any selinux keys?
</blockquote></details>

What is the mechanism to get selinux keys?

---

*[pkg/volume/flexvolume/mounter.go, line 80 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxJ-r1-80:-KdX4mgliLEumqkys5kk:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/mounter.go#L80)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary><blockquote>

if `fsGroup` is present and !readOnly, call [SetVolumeOwnership](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gce_pd/gce_pd.go#L311) 
</blockquote></details>

Done

---

*[pkg/volume/flexvolume/plugin.go, line 176 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxM-r1-176:-KdX5CHPc87zSpl-mnp8:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/plugin.go#L176)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary><blockquote>

->known
</blockquote></details>

Done

---

*[pkg/volume/flexvolume/util.go, line 2 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-2:-KdX5TclQ7G0W6BMzobi:bhnc2iu) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L2)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary><blockquote>

2017
</blockquote></details>

Done for all files.

---

*[pkg/volume/flexvolume/util.go, line 31 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-31:-KdX7BfcvHUgiA6eyiPw:b-b2tjc4) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L31)):*
> GetSecretForPV
Done

---

*[pkg/volume/flexvolume/util.go, line 66 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-66:-KdX7uJ7bpayD-BeaRtx:b1yesod) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L66)):*
>  named return but alreadyMounted and err are not used.
Fixed

---

*[pkg/volume/flexvolume/util.go, line 86 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-86:-KdXDz0-0xCarHYTCrPG:b-2e9uon) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L86)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary><blockquote>

what if the device needs a format? Can `FormatAndMount` be called?
</blockquote></details>

This call just bind mounts or mount an existing filesystem. So formatandmount is offloaded to the plugin using this interface. But FormatAndMount is also missing from MountDevice. Added it in mountDevice.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

chakri-nelluri commented:
Thanks @rootfs. Addressed the review comments. PTAL when you get a chance.

MikaelCluseau commented:
LGTM ;)

jsafrane commented:
Please add a release note. Follow the guide in new pr template at https://raw.githubusercontent.com/kubernetes/kubernetes/master/.github/PULL_REQUEST_TEMPLATE.md. This  ` ```release-note` section needs to be in the top comment of this PR.

chakri-nelluri commented:
Thanks @jsafrane. Will add it using the template.

---

Review status: 0 of 31 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/util/exec/exec.go, line 53 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXj-r2-53:-KdbHeGkhdtV91VQC5fP:baomdj4) ([raw file](https://github.com/kubernetes/kubernetes/blob/255b3f2eb1116291cd3bcc748235ccc4a1393b47/pkg/util/exec/exec.go#L53)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary><blockquote>

Looking at the implementation below, the process does not need to be stopped when this function returns, it can wait for SIGKILL. Put there at least a comment that it's expected behavior.
</blockquote></details>

Added the details.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 36 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r2-36:-KdbIS5QtUic-fePbqQ1:b-g3xo46) ([raw file](https://github.com/kubernetes/kubernetes/blob/255b3f2eb1116291cd3bcc748235ccc4a1393b47/pkg/volume/flexvolume/attacher-defaults.go#L36)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary><blockquote>

is this default attacher supposed to be used? If so, please lower the log level, warning is quite offensive here if it is part of normal operation. +update it in all default implementations below.

If it should not be used at all then return an error instead.
</blockquote></details>

Miss from debugging. Fixed the log level.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-:-KdbG9rtNUq5rJqagaCU:bdak4t)*
<!-- Sent from Reviewable.io -->

jsafrane commented on pkg/volume/flexvolume/plugin-defaults.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+func (plugin *pluginDefaults) GetVolumeName(spec *volume.Spec) (string, error) {
+	glog.Warning(logPrefix((*flexVolumePlugin)(plugin)), "using default GetVolumeName for volume ", spec.Name)
```
leftover warning

jsafrane commented:
the code looks solid to me, @rootfs, any comments?
@chakri-nelluri, don't forget to fix verification errors and unit tests, e.g. hack/verify-gofmt.sh, hack/verify-bazel.sh, make test, ...

rootfs commented:
@chakri-nelluri fix the CI errors then lgtm

chakri-nelluri commented:
Review status: 0 of 32 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/volume/flexvolume/plugin-defaults.go, line 32 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxL-r3-32:-KdgG2gqoSVb92GCxjoC:b-3kkq5i) ([raw file](https://github.com/kubernetes/kubernetes/blob/7979f5c6f6ee211fbb8973f3085e2f980c5237ae/pkg/volume/flexvolume/plugin-defaults.go#L32)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary><blockquote>

leftover warning
</blockquote></details>

This one I wanted it. GetVolumeName if we use default might not generate a unique name. So kept it a warning for debugging purposes.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

chakri-nelluri commented:
Thanks @rootfs @jsafrane. Fixed make test & verifiers.

saad-ali commented on pkg/volume/flexvolume/mounter-defaults.go:
```diff
@@ -0,0 +1,61 @@
+	return volume.Attributes{
+		ReadOnly:        f.readOnly,
+		Managed:         !f.readOnly,
+		SupportsSELinux: true,
```
Is this really true for all drivers?

saad-ali commented on pkg/volume/flexvolume/unmounter.go:
```diff
@@ -0,0 +1,80 @@
+
+func (f *flexVolumeUnmounter) TearDownAt(dir string) error {
+
+	if pathExists, pathErr := util.PathExists(dir); pathErr != nil {
```
Ditto the `UnmountDevice` command comment:  I suggest removing this logic. If a driver implements this call they can implement these checks. If they don't then the default `TearDownAt` command should just call UnmountPath(...) from pkg/volume/util/util.go which has the most robust implementation of this logic.

saad-ali commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -0,0 +1,191 @@
+func (plugin *flexVolumePlugin) unsupported(commands ...string) {
+	plugin.Lock()
+	defer plugin.Unlock()
+	plugin.unsupportedCommands = append(plugin.unsupportedCommands, commands...)
```
Trying to understand this. Is the purpose here to have a cache, so that you can fail fast based on previous invocations of driver, instead of having to invoke the driver and parse the response every time?

Also, does the `plugin` object  have a separate instance for every driver? How is that controlled?

saad-ali replied:
> Yes. It is for optimization.
> We have a separate plugin object for each driver.

Ack.

chakri-nelluri commented:
Review status: 0 of 33 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 36 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r4-36:-KdhCwbw30QgiEVku9P4:b-uxjw9k) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher-defaults.go#L36)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Make all the log messages for default methods `glog.Warningf` instead of `glog.V(5).Infof(`.
</blockquote></details>

We went the other way just to reduce the logging. Is it more for debugging? I am fine with either way. I will put them back.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 48 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r4-48:-KdhDajiRqY6S6MIw0Y2:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher-defaults.go#L48)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Print `mountsDir` in message as well.
</blockquote></details>

Done

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 59 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r4-59:-KdhDzRaAQPWDkJ43GhU:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher-defaults.go#L59)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Print `deviceMountPath` in message as well.
</blockquote></details>

Done

---

*[pkg/volume/flexvolume/attacher.go, line 106 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXr-r4-106:-KdhE-uCgrmCTKWnn3uH:b-hoz65z) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher.go#L106)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

`isAttached` should take the full list of volumes to check instead of issuing a separate `isAttached` call per volume. Since this call can happen frequently, we should avoid fanning out the number of backend calls (to prevent quota issues).
</blockquote></details>

I thought about it. The driver side callout implementation will be very tricky if we pass all specs at the same time on the command line. If we have the "unique volume name" passed to the call, we can use it. If not, we have to make another callout to get the "volume name", which is equally costly.

I agree with cost & quota issues. Any other way to do it?

---

*[pkg/volume/flexvolume/detacher.go, line 64 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxF-r4-64:-KdhMkGz4CBsMH9Utwbg:bls3lo) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/detacher.go#L64)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

I say just remove this logic. If a driver implements this call they can implement these checks. If they don't then the default unmount device command should just call `UnmountPath(...)` from `pkg/volume/util/util.go` which has the most robust implementation of this logic.
</blockquote></details>

That would be a behavior change with the existing drivers. Trying to reduce the behavior changes.

---

*[pkg/volume/flexvolume/detacher.go, line 94 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxF-r4-94:-KdhF1XuuRqHdl0kHMEn:bz7ui10) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/detacher.go#L94)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Is this check necessary, and is the message accurate?
</blockquote></details>

Os.Remove is failing if the path is removed by driver callout and old drivers used to do it.
Added the extra sanity check to reduce behavior changes with existing driver.

Message is not accurate (copy paste error). Fixed it.

---

*[pkg/volume/flexvolume/mounter-defaults.go, line 55 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxI-r4-55:-KdhGvaa4ittuT8DR8ii:b-6l9msz) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/mounter-defaults.go#L55)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Is this really true for all drivers?
</blockquote></details>

Not true for all the drivers. This is one of the things we wanted to fix with capabilities.

Returning true, as it is enforced/disabled at the OS level and ignored otherwise.

---

*[pkg/volume/flexvolume/plugin.go, line 173 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxM-r4-173:-KdhHWP7XOZuJwE8HY8g:bn9t6bv) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/plugin.go#L173)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Trying to understand this. Is the purpose here to have a cache, so that you can fail fast based on previous invocations of driver, instead of having to invoke the driver and parse the response every time?

Also, does the `plugin` object  have a separate instance for every driver? How is that controlled?
</blockquote></details>

Yes. It is for optimization. 
We have a separate plugin object for each driver.

---

*[pkg/volume/flexvolume/unmounter.go, line 47 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxQ-r4-47:-KdhMzBfP8a2cjIWRTj0:b-9panix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/unmounter.go#L47)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Ditto the `UnmountDevice` command comment:  I suggest removing this logic. If a driver implements this call they can implement these checks. If they don't then the default `TearDownAt` command should just call UnmountPath(...) from pkg/volume/util/util.go which has the most robust implementation of this logic.
</blockquote></details>

Same here. Trying to reduce behavior change.

---

*[pkg/volume/flexvolume/util.go, line 67 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r4-67:-KdhL3r77JwvyMbvIiu9:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/util.go#L67)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

`IsLikelyNotMountPoint(...)` inside `isNotMounted(...)` does a `os.Stat(...)` and returns the result. Instead calling it twice just use the result from that call. Example: https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gce_pd/attacher.go#L188
</blockquote></details>

Done

---

*[pkg/volume/flexvolume/util.go, line 105 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r4-105:-KdhLH7cnQGgGCSKhq3O:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/util.go#L105)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

Remove this. Use `UnmountPath(...)` from `pkg/volume/util/util.go` instead.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

chakri-nelluri commented:
Thanks @saad-ali. Addressed review comments. PTAL

chakri-nelluri commented:
Review status: 0 of 33 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/util/exec/exec.go, line 53 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXj-r2-53:-KdhhAT--aHWSUSmjuAp:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/255b3f2eb1116291cd3bcc748235ccc4a1393b47/pkg/util/exec/exec.go#L53)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Added the details.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 36 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r2-36:-KdhhD7D4SHN1Df0S32A:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/255b3f2eb1116291cd3bcc748235ccc4a1393b47/pkg/volume/flexvolume/attacher-defaults.go#L36)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Miss from debugging. Fixed the log level.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 36 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r4-36:-KdhhDMazjqZWGEk4_91:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher-defaults.go#L36)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

We went the other way just to reduce the logging. Is it more for debugging? I am fine with either way. I will put them back.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 48 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r4-48:-KdhhIjLd3tFGX33Znju:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher-defaults.go#L48)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 59 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r4-59:-KdhhHyOMMspru4baoJQ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher-defaults.go#L59)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/detacher.go, line 64 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxF-r4-64:-Kdhhd4ZZ_EFCPpNkaic:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/detacher.go#L64)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

That would be a behavior change with the existing drivers. Trying to reduce the behavior changes.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/detacher.go, line 94 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxF-r4-94:-KdhhcUrk90V2aKFlYsb:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/detacher.go#L94)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Os.Remove is failing if the path is removed by driver callout and old drivers used to do it.
Added the extra sanity check to reduce behavior changes with existing driver.

Message is not accurate (copy paste error). Fixed it.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/driver-call.go, line 51 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxH-r1-51:-KdhhlvvwlI9dAmFBeY4:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/driver-call.go#L51)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

What is the mechanism to get selinux keys?
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/mounter-defaults.go, line 55 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxI-r4-55:-KdhhUO2RC_MLJgxp2xI:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/mounter-defaults.go#L55)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Not true for all the drivers. This is one of the things we wanted to fix with capabilities.

Returning true, as it is enforced/disabled at the OS level and ignored otherwise.

</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/mounter.go, line 80 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxJ-r1-80:-KdhhV4pslQ3Bb6YgM89:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/mounter.go#L80)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/plugin-defaults.go, line 32 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxL-r3-32:-KdhhWCSEnvJ6qz95lxi:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/229499412dd9e7d3260dd7cbbcfc02fa8bcf2471/pkg/volume/flexvolume/plugin-defaults.go#L32)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

This one I wanted it. GetVolumeName if we use default might not generate a unique name. So kept it a warning for debugging purposes.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/plugin.go, line 176 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxM-r1-176:-KdhhXU2ecfoDTD8jISj:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/plugin.go#L176)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/plugin.go, line 173 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxM-r4-173:-KdhhWun4yK2TES-9Wue:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/plugin.go#L173)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Yes. It is for optimization. 
We have a separate plugin object for each driver.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/unmounter.go, line 47 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxQ-r4-47:-Kdhh_dkEeJ8W8zW7cNj:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/unmounter.go#L47)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Same here. Trying to reduce behavior change.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 2 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-2:-Kdhhf4qX9jiPyzq0O7C:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L2)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done for all files.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 31 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-31:-Kdhhg3Wsej5lihu67BZ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L31)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

> GetSecretForPV
Done
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 66 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-66:-KdhhhqPaswvsJ7ukgwK:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L66)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

>  named return but alreadyMounted and err are not used.
Fixed
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 86 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-86:-KdhhjmPxH3PiIK8Z4xp:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L86)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

This call just bind mounts or mount an existing filesystem. So formatandmount is offloaded to the plugin using this interface. But FormatAndMount is also missing from MountDevice. Added it in mountDevice.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 67 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r4-67:-KdhhiZ8H-FGlYJpy2M9:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/util.go#L67)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 105 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r4-105:-KdhhkChB4uZw0QeFZ2O:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/util.go#L105)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

chakri-nelluri commented:
Review status: 0 of 33 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/volume/flexvolume/util.go, line 2 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-2:-Kdhhw_FHiVrTStt-E4m:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L2)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 31 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-31:-KdhhwEJrhBB06CzY6B0:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L31)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 66 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-66:-Kdhhvid-jH3Lxbt-I-I:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L66)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 86 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-86:-Kdhhu9ztMfM5oKd9ssq:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L86)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 67 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r4-67:-Kdhhv_2HhmzTzH-mb1l:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/util.go#L67)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

chakri-nelluri commented:
Review status: 0 of 33 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/volume/flexvolume/util.go, line 31 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-31:-KdhhxiHr7MZW_1RlUk3:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L31)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 66 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-66:-Kdhhy33mtzDNU53TusT:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L66)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 86 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r1-86:-KdhhyPhGOQS5y1vg1s2:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bdff3b27c13982885fcbc75c6db715e529fb28e7/pkg/volume/flexvolume/util.go#L86)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[pkg/volume/flexvolume/util.go, line 105 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxS-r4-105:-KdhhykQV6EUOTXgZPw8:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/util.go#L105)):*
<details><summary><i>Previously, chakri-nelluri (Chakravarthy Nelluri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

MikaelCluseau commented on pkg/volume/flexvolume/attacher.go:
```diff
@@ -0,0 +1,120 @@
+		volumesAttachedCheck[spec] = true
+
+		call := a.plugin.NewDriverCall(isAttached)
+		call.AppendSpec(spec, a.plugin.host, nil)
```
I should probably be called "areAttached" then, too?

saad-ali replied:
Saw your comment:

> I thought about it. The driver side callout implementation will be very tricky if we pass all specs at the same time on the command line. If we have the "unique volume name" passed to the call, we can use it. If not, we have to make another callout to get the "volume name", which is equally costly.
> 
> I agree with cost & quota issues. Any other way to do it?

Processing

saad-ali replied:
I'll let you and @MikaelCluseau decide, if it is too difficult to implement.

chakri-nelluri commented:
Review status: 0 of 33 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[pkg/volume/flexvolume/attacher.go, line 106 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXr-r4-106:-KdlXVtoqCH62NlKo8nS:b8wksqb) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher.go#L106)):*
<details><summary><i>Previously, MikaelCluseau (Mikaël Cluseau) wrote…</i></summary><blockquote>

I should probably be called "areAttached" then, too?
</blockquote></details>

Yep. If we change it.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

saad-ali commented on pkg/volume/flexvolume/detacher.go:
```diff
@@ -0,0 +1,98 @@
+	if pathExists, pathErr := util.PathExists(deviceMountPath); pathErr != nil {
+		return fmt.Errorf("Error checking if path exists: %v", pathErr)
+	} else if !pathExists {
+		glog.Warningf("Warning: Unmount skipped because path does not exist: %v", deviceMountPath)
```
nit: @chakri-nelluri Message should be something like "Skipped removing directory because directory does not exist" (the lack of directory does not indicate that unmount was skipped) and probably an Info is sufficient instead of warning.

saad-ali commented:
Almost there. A couple of comments. Once those are addressed, please squash commits, and modify the release-note to be more descriptive. Indicate that Flex volume driver interface is being modified to expose new attach/detach hooks, that these are different from previous releases because... and they will require Flex Driver authors to update their code accordingly.

chakri-nelluri commented:
Review status: 0 of 33 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[pkg/volume/flexvolume/attacher-defaults.go, line 36 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXq-r2-36:-Kdmpid-fQUAAMt71BJL:bb6sh2i) ([raw file](https://github.com/kubernetes/kubernetes/blob/255b3f2eb1116291cd3bcc748235ccc4a1393b47/pkg/volume/flexvolume/attacher-defaults.go#L36)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

`Warning` is desired because of the way that Flex is implemented. It implements the attacher interface but individual drivers may or may not implement attacher methods. Therefore, we want it to be very clear when Flex decides to use a default method.
</blockquote></details>

Ack.

---

*[pkg/volume/flexvolume/attacher.go, line 106 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFPdSLKEQh4DAXr-r4-106:-KdmofQ4lE1PfdpN2nfg:bniroxb) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/attacher.go#L106)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

I'll let you and @MikaelCluseau decide, if it is too difficult to implement.
</blockquote></details>

It is too difficult to implement for driver writers. As per isAttached vs areAttached, I would keep it isAttached as it is a per volume call out. :)

---

*[pkg/volume/flexvolume/detacher.go, line 94 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804#-KdVzHFQZceyjDyLETxF-r4-94:-KdmpL1bf3BybfKtrsAU:bx6fbdl) ([raw file](https://github.com/kubernetes/kubernetes/blob/9279c01f3e18de7d4979c26657a9e551d021b668/pkg/volume/flexvolume/detacher.go#L94)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

nit: @chakri-nelluri Message should be something like "Skipped removing directory because directory does not exist" (the lack of directory does not indicate that unmount was skipped) and probably an Info is sufficient instead of warning.
</blockquote></details>

Oops. Somehow this change is missing. I removed the log altogether.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41804)*
<!-- Sent from Reviewable.io -->

chakri-nelluri commented:
Thanks @saad-ali. Addressed the review comments.

saad-ali commented:
/lgtm
/approve

saad-ali commented:
Please remember to: 

> Modify the release-note to be more descriptive. Indicate that Flex volume driver interface is being modified to expose new attach/detach hooks, that these are different from previous releases because... and they will require Flex Driver authors to update their code accordingly.

chakri-nelluri commented:
Thanks @saad-ali and
of course @MikaelCluseau most of the code changes are from his previous PR.

klausenbusk commented:
It seems like the (README)[https://github.com/kubernetes/kubernetes/tree/master/examples/volumes/flexvolume] was forgotten in this PR, it still show the old design.

chakri-nelluri commented:
@klausenbusk I am working on updating the documentation. It should be in tomorrow.

klausenbusk commented:
> @klausenbusk I am working on updating the documentation. It should be in tomorrow.

That sounds great :) I'm properly going to update https://github.com/tonyzou/flexvolumes DigitalOcean plugin for 1.6 support.

klausenbusk commented:
Another question, shouldn't we also pass secret to the detach function? or should the API key be stored outside k8s?
This is properly a new issue, I'm just wondering if this is intentional or?

MikaelCluseau commented:
@klausenbusk secrets are namespaced while volumes aren't so, with the redesign, it was chosen to drop this feature. Some background is here: https://github.com/kubernetes/kubernetes/pull/26926#issuecomment-224291589.

chakri-nelluri commented:
@klausenbusk Updated documentation is at https://github.com/kubernetes/community/tree/master/contributors/devel/flexvolume.md.

Ping me if you have any questions.

klausenbusk commented:
> Ping me if you have any questions.

I couldn't get my K8s 1.6 cluster to work (kube-dns issues), so I'm "stuck" on 1.5.4 for a little while (properly until the CoreOS guys update their [scripts](https://github.com/coreos/coreos-kubernetes)).

I have been wondering about one thing.
My k8s cluster is running on DigitalOcean, and they "only" allow 5 "block storage disk" per node. Can the current k8s flexvolume code handle that somehow? Ex. can my flexvolume plugin say "Disk limit reached, please find another node"? I'm properly gonna open a new issue for that, I just wanna hear what you think, before I do that.

I think it is very unlikely that I end in a situation where 5 disk is attached to a single droplet, but someone is properly gonna hit the limit at some point.

MikaelCluseau commented:
@klausenbusk I think it's more a scheduler issue; I think I've seen something about that for GCE or AWS limit at 16 volumes/node.

MikaelCluseau commented:
yep, it's #7835 -> #19880 ; I think you can open an issue about the DigitalOcean case, but it's not the responsibility of the volume plugin so there's no interface to say that (for now at least).

klausenbusk commented:
> yep, it's #7835 -> #19880 ; I think you can open an issue about the DigitalOcean case, but it's not the responsibility of the volume plugin so there's no interface to say that (for now at least).

Thanks, I did a little more searching and stumbled on: https://github.com/kubernetes/kubernetes/issues/24317 which links to https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#opaque-integer-resources-alpha-feature  which could solve the issue. There is also `KUBE_MAX_PD_VOLS`, I'm not exactly sure how it works, if I set it to 5 will it work then?

chakri-nelluri commented:
@klausenbusk Check if you can leverage https://github.com/kubernetes/kubernetes/pull/13580 to integrate with. There is a new bind proposal for scheduler https://github.com/kubernetes/kubernetes/pull/41447, trying to address the limitations with 13580.

klausenbusk commented:
> @klausenbusk Check if you can leverage #13580 to integrate with. There is a new bind proposal for scheduler #41447, trying to address the limitations.

I'm wondering if it would make sense to extend the `MaxPDVolumeCountChecker` to support flexvolume plugins?
